noxfile.py

```diff
@@ -0,0 +1,170 @@
+"""
+Nox is a command-line tool that automates testing in multiple Python environments, similar to tox.
+Unlike tox, nox uses a standard Python file for configuration.
+
+To run all sessions defined in the noxfile.py file, run the following command:
+$ nox
+
+To run a specific session, run the following command:
+$ nox -s <session-name>
+
+To run a session with additional arguments, run the following command:
+$ nox -s <session-name> -- <additional-arguments>
+
+To list all available sessions, run the following command:
+$ nox -l
+"""
+
+import os
+from pathlib import Path
+
+import nox
+import tomli
+from nox.sessions import Session
+
+# default sessions (sorted alphabetically)
+nox.options.sessions = ["bandit", "black", "flake8", "isort", "mypy", "unit-tests"]
+
+# reuse virtual environment across all sessions
+nox.options.reuse_venv = "always"
+
+
+def install_requirements(session: Session) -> None:
+    """Install requirements for all sessions"""
+    session.install("-r", "requirements-extras.txt")
+
+
+def parse_supported_python_versions() -> list[str]:
+    """Parse supported Python versions from pyproject.toml"""
+    pyproject_path = Path("pyproject.toml")
+    pyproject = tomli.loads(pyproject_path.read_text())
+    classifiers: list[str] = pyproject["project"]["classifiers"]
+
+    return [
+        c.split(" :: ")[-1]
+        for c in classifiers
+        if c.startswith("Programming Language :: Python :: 3") and c.count(".") == 1
+    ]
+
+
+@nox.session()
+def bandit(session: Session) -> None:
+    """Run bandit on cachi2 directory"""
+    install_requirements(session)
+    cmd = "bandit -c pyproject.toml -r cachi2"
+    session.run(*cmd.split(), *session.posargs, silent=True)
+
+
+@nox.session()
+def black(session: Session) -> None:
+    """Run black on cachi2 and tests directories and noxfile.py"""
+    install_requirements(session)
+    cmd = "black --check --diff cachi2 tests noxfile.py"
+    session.run(*cmd.split(), *session.posargs, silent=True)
+
+
+@nox.session()
+def flake8(session: Session) -> None:
+    """Run flake8 on cachi2 and tests directories"""
+    install_requirements(session)
+    cmd = "flake8 cachi2 tests"
+    session.run(*cmd.split(), *session.posargs, silent=True)
+
+
+@nox.session()
+def isort(session: Session) -> None:
+    """Run isort on cachi2 and tests directories and noxfile.py"""
+    install_requirements(session)
+    cmd = "isort --check --diff --color cachi2 tests noxfile.py"
+    session.run(*cmd.split(), *session.posargs, silent=True)
+
+
+@nox.session()
+def mypy(session: Session) -> None:
+    """Run mypy on cachi2 and tests directories"""
+    install_requirements(session)
+    cmd = "mypy --install-types --non-interactive cachi2 tests"
+    session.run(*cmd.split(), *session.posargs, silent=True)
```

Comment:
Obligatory [non-blocking] tag.

Would this be too much of generalization?

```python
@nox.session()
def mypy(session: Session) -> None:
    """Run mypy on cachi2 and tests directories"""
    run_cmd_for(session, "mypy --install-types --non-interactive cachi2 tests")

def run_cmd_for(session, cmd, loud=False):
    install_requirements(session)
    session.run(*cmd.split(), *session.posargs, silent=not loud)

Reply:
I don't know, to me it feels like an unnecessary wrapper around each command.

Reply:
Given that all of the commands are exactly the same, maintainability would be increased by

```
def run_cmd(cmd, session):
    session.run(*cmd.split(), *session.posargs, silent=True)
```

called from the individual sessions as `run_cmd(cmd, session)`.

Why? I realize that it seems excessive, but:

- It actually reduces the amount of repetition. If we need to change the  options to `session.run`, we can do it in ONE place instead of ten
- If we do have one specific session which needs to have a slightly different `session.run`, we can call `session.run` directly *from that specific session*, i.e. ignoring our "wrapper". As it stands, if one out the many sessions has different options, it's really hard to see.